src/main/java/com/uid2/shared/auth/AuthorizableStore.java

```diff
@@ -1,77 +1,153 @@
-
-        public AuthorizableStoreSnapshot(Collection<T> authorizables) {
-            this.authorizables = authorizables.stream()
-                    .collect(Collectors.toMap(
-                            a -> ByteBuffer.wrap(Base64.getDecoder().decode(a.getKeyHash())),
-                            a -> a
-                    ));
-            this.salts = authorizables.stream()
-                    .map(a -> Base64.getDecoder().decode(a.getKeySalt()))
-                    .collect(Collectors.toList());
-        }
-
-        public Map<ByteBuffer, T> getAuthorizables() {
-            return authorizables;
-        }
-
-        public Collection<byte[]> getSalts() {
-            return salts;
-        }
-    }
-
-    private static final KeyHasher KEY_HASHER = new KeyHasher();
     private static final Logger LOGGER = LoggerFactory.getLogger(AuthorizableStore.class);
+    private static final Pattern KEY_PATTERN = Pattern.compile("(?:UID2|EUID)-[CO]-[LTIP]-([0-9]+)-.{6}\\..{38}");
+    private static final KeyHasher KEY_HASHER = new KeyHasher();
+    private static final int CACHE_MAX_SIZE = 100_000;
 
     private final AtomicReference<AuthorizableStoreSnapshot> authorizables;
+    private final LoadingCache<String, String> keyToHashCache;
 
     public AuthorizableStore() {
         this.authorizables = new AtomicReference<>(new AuthorizableStoreSnapshot(new ArrayList<>()));
+        this.keyToHashCache = createCache();
     }
 
     public void refresh(Collection<T> authorizablesToRefresh) {
         authorizables.set(new AuthorizableStoreSnapshot(authorizablesToRefresh));
     }
 
-    public T getFromKey(String key) {
+    public T getAuthorizableByKey(String key) {
         AuthorizableStoreSnapshot latest = authorizables.get();
-        // TODO: add caching
-        for (byte[] salt : latest.getSalts()) {
+
+        T authorizable;
+
+        authorizable = getAuthorizableBySiteId(latest, key);
+        if (authorizable != null) {
+            return authorizable;
+        }
+
+        String cachedHash = keyToHashCache.get(key);
+        if (cachedHash != null) {
+            return cachedHash.isBlank() ? null : latest.getAuthorizableByHash(wrapHashToByteBuffer(cachedHash));
+        }
+
+        authorizable = getAuthorizableIfKeyExists(latest, key);
+        keyToHashCache.put(key, authorizable == null ? "" : authorizable.getKeyHash());
+
+        return authorizable;
+    }
+
+    public T getAuthorizableByHash(String hash) {
+        ByteBuffer hashBytes = wrapHashToByteBuffer(hash);
+        if (hashBytes == null) {
+            return null;
+        }
+
+        AuthorizableStoreSnapshot latest = authorizables.get();
+        return latest.getAuthorizableByHash(hashBytes);
+    }
+
+    private T getAuthorizableBySiteId(AuthorizableStoreSnapshot snapshot, String key) {
+        Integer siteId = getSiteIdFromKey(key);
+
+        if (siteId != null) {
+            T authorizable = snapshot.getAuthorizableBySiteId(siteId);
```

Comment:
Can there be multiple authorizables per site ID? I think so

Reply:
Should probably add support for it and maybe these should be cached. Perhaps the site ID only reduces the number of salts to search, might make the code cleaner